mount/mount_linux.go

```diff
@@ -292,18 +297,71 @@ func (mounter *SafeFormatAndMount) checkAndRepairXfsFilesystem(source string) er
 			return nil
 		} else {
 			klog.Warningf("Filesystem corruption was detected for %s, running xfs_repair to repair", source)
-			out, err := mounter.Exec.Command("xfs_repair", args...).CombinedOutput()
-			if err != nil {
-				return NewMountError(HasFilesystemErrors, "'xfs_repair' found errors on device %s but could not correct them: %s\n", source, out)
-			} else {
-				klog.Infof("Device %s has errors which were corrected by xfs_repair.", source)
-				return nil
+			var uncorrectedErr error
+
+			// If xfs_repair fails to repair the file system successfully, try giving the same xfs_repair
+			// command twice more; xfs_repair may be able to make more repairs on successive runs.
+			for i := 0; i < 3; i++ {
+				out, err := mounter.Exec.Command("xfs_repair", args...).CombinedOutput()
+				if err != nil {
+					if e, isExitError := err.(utilexec.ExitError); isExitError {
+						switch e.ExitStatus() {
+						case xfsRepairErrorsUncorrected:
+							uncorrectedErr = NewMountError(HasFilesystemErrors, "'xfs_repair' found errors on device %s but could not correct them: %s\n", source, string(out))
+						// If the exit status is 2, do not retry, replay the dirty logs instead.
+						case xfsRepairErrorsDirtyLogs:
+							return mounter.replayXfsDirtyLogs(source)
+						default:
+							return NewMountError(HasFilesystemErrors, "'xfs_repair' found errors on device %s but could not correct them: %s\n", source, string(out))
+						}
+					} else {
+						return NewMountError(HasFilesystemErrors, "failed to run 'xfs_repair' on device %s: %v\n", source, err)
+					}
+				} else {
+					klog.Infof("Device %s has errors which were corrected by xfs_repair.", source)
+					return nil
```

Comment:
nit - Can we flatten some of this nested by returning early?

```golang
out, err := mounter.Exec.Command("xfs_repair", args...).CombinedOutput()
	if err == nil {
		klog.Infof("Device %s has errors which were corrected by xfs_repair.", source)
		return nil
	}
	e, isExitError := err.(utilexec.ExitError)
	if !isExitError {
		return NewMountError(HasFilesystemErrors, "failed to run 'xfs_repair' on device %s: %v\n", source, err)
		
	}
	switch e.ExitStatus() {
	case xfsRepairErrorsUncorrected:
		uncorrectedErr = NewMountError(HasFilesystemErrors, "'xfs_repair' found errors on device %s but could not correct them: %s\n", source, string(out))
		// If the exit status is 2, do not retry, replay the dirty logs instead.
	case xfsRepairErrorsDirtyLogs:
		return mounter.replayXfsDirtyLogs(source)
	default:
		return NewMountError(HasFilesystemErrors, "'xfs_repair' found errors on device %s but could not correct them: %s\n", source, string(out))
	}
```

Reply:
fixed, it looks better, thanks.